### DISCUSSION TITLE
METplus wrapper return codes

### DISCUSSION THREAD

**Discussion by gsketefian** [dtcenter/METplus#1731] [Statistical Computation] [answered]
Hi,

I have a bash script that calls `run_metplus.py` and passes it a couple of configuration files, e.g.:
```
${METPLUS_PATH}/ush/run_metplus.py \
  -c ${METPLUS_CONF}/common.conf \
  -c ${METPLUS_CONF}/GridStat_APCP01h.conf
```
This script is part of a rocoto workflow, and the workflow determines whether the script succeeded or failed depending on the script's return value.  I would like to base that on the return value from `run_metplus.py` (which I assume is based on the return value from the specific wrapper that it calls, e.g. the one for `gen_ens_prod`), but in the METplus documentation I could not find a list of return values.  Could you please point me to where I can find that information?

Thanks,
Gerard

**Comment by JohnHalleyGotway** [comment]
Gerard, I agree that I do not see a description of return METplus wrapper return codes in the documentation.

I've talked to @georgemccabe offline and have asked him to comment about this here.

**Comment by georgemccabe** [comment]
@gsketefian,

run_metplus.py will return zero if everything runs successfully. If there are any errors in the run, then a non-zero value is returned. The return value is almost always 1 on error, but it looks like 2 is returned on some occasions such as an incorrect argument provided to the script. I agree that it would be good to document the return values.

**Comment by gsketefian** [comment]
Thanks @georgemccabe.

I guess the question then is, what is considered an error?  If I'm verifying a bunch of dates, say 30, and on 2 of those dates there were missing obs for 3 of say 20 variables I'm verifying, would that cause `run_metplus.py` to return a non-zero value?

From the point of view of running a workflow, I would not consider this scenario a failure/error since very likely the obs files exist but are missing some of the needed fields at certain times.  Rerunning the vx task within the workflow wouldn't change the result.  

On the other hand, if the obs file itself doesn't exist, if it doesn't have read permission for the user, etc, those I would want to be reported back to rocoto as errors since these are things the user can probably fix and then rerun the vx task.

I'm trying to distinguish between these different scenarios so I can have the calling script take the action that makes the most sense.  I was hoping different return values would indicate different verification end states.

**Reply by JohnHalleyGotway** [reply]
@gsketefian when running the Point-Stat tool in MET, finding 0 observations matching the forecast data is NOT an error. In this way you can run Point-Stat and produce 0 matched pairs across all the variables being evaluated and it will still return good status (i.e. return code = 0). However, if you tell Point-Stat to read a forecast or observation file name that does not actually exist, it will error out with bad status. So the Point-Stat tool is behaving the way you expect.

@georgemccabe can you clarify the behavior of the METplus wrapper? Let's say we've configured it to run Point-Stat every 6 hours to verify using GDAS point observations. But the GDAS point observation file does not exist for one of the times. Does that trigger METplus to throw an error? Or does that result in a warning?

**Comment by georgemccabe** [comment, marked as answer]
If an expected input file does not exist, the then wrappers will report an error and continue to the next run time, which will result in a non-zero return value at the end of execution. If the expected input files do exist, then the actual MET executable will determine if an error should be reported or not. In the example @JohnHalleyGotway provided, 0 observations in the input will not result in an error.

**Comment by gsketefian** [comment]
@JohnHalleyGotway @georgemccabe Thanks for your answers, it's a lot clearer now.  I'll mark @georgemccabe's as the answer since it's about the wrapper return values.  It would be nice to have something about return values in the documentation.
